server/src/main/java/org/elasticsearch/transport/RemoteClusterService.java

```diff
@@ -168,6 +164,15 @@ private synchronized void updateRemoteClusters(Map<String, List<DiscoveryNode>>
         this.remoteClusters = Collections.unmodifiableMap(remoteClusters);
     }
 
+    static Predicate<DiscoveryNode> getNodePredicate(Settings settings) {
+        if (REMOTE_NODE_ATTRIBUTE.exists(settings)) {
+            // nodes can be tagged with node.attr.remote_gateway: true to allow a node to be a gateway node for cross cluster search
+            String attribute = REMOTE_NODE_ATTRIBUTE.get(settings);
+            return DEFAULT_NODE_PREDICATE.and((node) -> Booleans.parseBoolean(node.getAttributes().getOrDefault(attribute, "false")));
+        }
+        return DEFAULT_NODE_PREDICATE;
```

Comment:
question: when node attributes are configured, should we still apply the selection based on node roles? I am doing so, but I want to double check. With this change there is no way to ever have a dedicated master act as a gateway, even if node attributes are configured to do so. I think that's a good thing though.

I am not sure whether we should have some protection for cases where we connect to a cluster that has only dedicated masters on compatible versions. That should be already in place as it could happen before too that no nodes are left after the selection.

Reply:
I think so too!